File: plugin-core/src/main/java/com/github/catatafishen/ideagentforcopilot/settings/ChatHistoryConfigurable.java
package com.github.catatafishen.ideagentforcopilot.settings;

import com.google.gson.JsonArray;
import com.google.gson.JsonParser;
import com.intellij.icons.AllIcons;
import com.intellij.ide.actions.RevealFileAction;
import com.intellij.notification.NotificationGroupManager;
import com.intellij.notification.NotificationType;
import com.intellij.openapi.actionSystem.ActionUpdateThread;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.Messages;
import com.intellij.ui.ColoredTableCellRenderer;
import com.intellij.ui.SimpleTextAttributes;
import com.intellij.ui.ToolbarDecorator;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.table.JBTable;
import com.intellij.util.ui.FormBuilder;
import com.intellij.util.ui.JBUI;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.table.AbstractTableModel;
import javax.swing.table.DefaultTableCellRenderer;
import java.io.IOException;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Project-level settings page for managing chat history files stored on disk.
 * Appears under Settings → AgentBridge → Other → Chat History.
 * <p>
 * Deletions happen immediately via toolbar actions (like JetBrains' own cache cleanup pages),
 * so {@link #isModified()} always returns {@code false} and {@link #apply()} is a no-op.
 */
public final class ChatHistoryConfigurable implements Configurable {

    public static final String ID = "com.github.catatafishen.ideagentforcopilot.chatHistory";

    private static final Logger LOG = Logger.getInstance(ChatHistoryConfigurable.class);

    private static final String AGENT_WORK_DIR = ".agent-work";
    private static final String CONVERSATIONS_DIR = "conversations";
    private static final String CURRENT_SESSION_FILE = "conversation.json";
    private static final String ARCHIVE_PREFIX = "conversation-";
    private static final String JSON_EXTENSION = ".json";

    private static final DateTimeFormatter TIMESTAMP_PARSER =
        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH-mm-ss");
    private static final DateTimeFormatter DISPLAY_FORMATTER =
        DateTimeFormatter.ofPattern("MMM d, yyyy h:mm a");

    private static final String CURRENT_SESSION_LABEL = "Current Session";

    private static final long KB = 1024L;
    private static final long MB = KB * 1024L;

    private final Project project;

    private JBLabel summaryLabel;
    private JBTable table;
    private ConversationTableModel tableModel;

    public ChatHistoryConfigurable(Project project) {
        this.project = project;
    }

    @Override
    public @Nls(capitalization = Nls.Capitalization.Title) String getDisplayName() {
        return "Chat History";
    }

    @Override
    public @Nullable JComponent createComponent() {
        summaryLabel = new JBLabel();
        summaryLabel.setBorder(JBUI.Borders.empty(0, 0, 4, 0));

        tableModel = new ConversationTableModel();
        table = new JBTable(tableModel);
        table.setSelectionMode(ListSelectionModel.MULTIPLE_INTERVAL_SELECTION);
        table.getEmptyText().setText("No conversations found");
        table.getEmptyText().appendSecondaryText(
            "Start a chat to create conversation history",
            SimpleTextAttributes.GRAYED_ATTRIBUTES, null);
        table.getAccessibleContext().setAccessibleName("Conversation history files");

        table.getColumnModel().getColumn(0).setPreferredWidth(JBUI.scale(200));
        table.getColumnModel().getColumn(1).setPreferredWidth(JBUI.scale(80));
        table.getColumnModel().getColumn(2).setPreferredWidth(JBUI.scale(80));
        table.getColumnModel().getColumn(3).setPreferredWidth(JBUI.scale(160));

        table.getColumnModel().getColumn(0).setCellRenderer(new ConversationNameRenderer());

        DefaultTableCellRenderer rightAligned = new DefaultTableCellRenderer();
        rightAligned.setHorizontalAlignment(SwingConstants.RIGHT);
        table.getColumnModel().getColumn(1).setCellRenderer(new MessageCountRenderer());
        table.getColumnModel().getColumn(2).setCellRenderer(rightAligned);

        JPanel decorated = ToolbarDecorator.createDecorator(table)
            .disableAddAction()
            .disableUpDownActions()
            .setRemoveAction(b -> deleteSelectedConversations())
            .setRemoveActionUpdater(e -> {
                int[] rows = table.getSelectedRows();
                if (rows.length == 0) return false;
                for (int row : rows) {
                    if (tableModel.getEntryAt(row).isCurrentSession()) return false;
                }
                return true;
            })
            .addExtraAction(createDeleteAllArchivesAction())
            .addExtraAction(createRefreshAction())
            .addExtraAction(createRevealInFinderAction())
            .createPanel();

        loadConversations();

        JPanel panel = FormBuilder.createFormBuilder()
            .addComponent(new JBLabel(
                "<html>Conversation files stored in this project's "
                    + "<code>.agent-work</code> directory.</html>"))
            .addComponent(summaryLabel)
            .addComponentFillVertically(decorated, 0)
            .getPanel();
        panel.setBorder(JBUI.Borders.empty(8));
        return panel;
    }

    @Override
    public boolean isModified() {
        return false;
    }

    @Override
    public void apply() {
        // Deletions happen immediately — nothing to apply.
    }

    @Override
    public void reset() {
        loadConversations();
    }

    @Override
    public void disposeUIResources() {
        summaryLabel = null;
        table = null;
        tableModel = null;
    }

    // ── Toolbar actions ──

    private AnAction createDeleteAllArchivesAction() {
        return new AnAction("Delete All Archives", "Delete all archived conversations (keeps current session)",
            AllIcons.Actions.GC) {
            @Override
            public void actionPerformed(@NotNull AnActionEvent e) {
                int result = Messages.showYesNoDialog(
                    project,
                    "Delete all archived conversations?\nThe current session will be kept.\n\nThis action cannot be undone.",
                    "Delete All Archives",
                    Messages.getWarningIcon());
                if (result != Messages.YES) return;

                List<ConversationEntry> entries = tableModel.getEntries();
                List<Path> toDelete = new ArrayList<>();
                for (ConversationEntry entry : entries) {
                    if (!entry.isCurrentSession()) {
                        toDelete.add(entry.path());
                    }
                }
                deleteFiles(toDelete);
                loadConversations();
            }

            @Override
            public @NotNull ActionUpdateThread getActionUpdateThread() {
                return ActionUpdateThread.EDT;
            }

            @Override
            public void update(@NotNull AnActionEvent e) {
                boolean hasArchives = tableModel != null
                    && tableModel.getEntries().stream().anyMatch(entry -> !entry.isCurrentSession());
                e.getPresentation().setEnabled(hasArchives);
            }
        };
    }

    private AnAction createRefreshAction() {
        return new AnAction("Refresh", "Rescan conversation directory", AllIcons.Actions.Refresh) {
            @Override
            public void actionPerformed(@NotNull AnActionEvent e) {
                loadConversations();
            }

            @Override
            public @NotNull ActionUpdateThread getActionUpdateThread() {
                return ActionUpdateThread.EDT;
            }
        };
    }

    private AnAction createRevealInFinderAction() {
        return new AnAction("Show in " + RevealFileAction.getFileManagerName(),
            "Open conversations directory in file manager",
            AllIcons.Actions.MenuOpen) {
            @Override
            public void actionPerformed(@NotNull AnActionEvent e) {
                String basePath = project.getBasePath();
                if (basePath == null) return;
                Path dir = Path.of(basePath, AGENT_WORK_DIR, CONVERSATIONS_DIR);
                if (!Files.isDirectory(dir)) {
                    dir = Path.of(basePath, AGENT_WORK_DIR);
                }
                RevealFileAction.openDirectory(dir.toFile());
            }

            @Override
            public @NotNull ActionUpdateThread getActionUpdateThread() {
                return ActionUpdateThread.EDT;
            }
        };
    }

    private void deleteSelectedConversations() {
        int[] selectedRows = table.getSelectedRows();
        if (selectedRows.length == 0) return;

        String message = selectedRows.length == 1
            ? "Delete the selected conversation?\n\nThis action cannot be undone."
            : "Delete " + selectedRows.length + " selected conversations?\n\nThis action cannot be undone.";

        int result = Messages.showYesNoDialog(project, message, "Delete Conversations", Messages.getWarningIcon());
        if (result != Messages.YES) return;

        List<Path> toDelete = new ArrayList<>();
        for (int row : selectedRows) {
            toDelete.add(tableModel.getEntryAt(row).path());
        }
        deleteFiles(toDelete);
        loadConversations();
    }

    private void deleteFiles(List<Path> paths) {
        List<String> failures = new ArrayList<>();
        for (Path path : paths) {
            try {
                Files.deleteIfExists(path);
            } catch (IOException e) {
                LOG.warn("Failed to delete conversation file: " + path, e);
                failures.add(path.getFileName().toString());
            }
        }
        if (!failures.isEmpty()) {
            NotificationGroupManager.getInstance()
                .getNotificationGroup("AgentBridge Notifications")
                .createNotification(
                    "Failed to delete " + failures.size() + " file(s): " + String.join(", ", failures),
                    NotificationType.WARNING)
                .notify(project);
        }
    }

    // ── Data loading ──

    private void loadConversations() {
        if (tableModel == null) return;

        table.getEmptyText().setText("Loading\u2026");
        tableModel.setEntries(List.of());
        updateSummary(List.of());

        ApplicationManager.getApplication().executeOnPooledThread(() -> {
            List<ConversationEntry> entries = scanConversations();
            entries.sort(Comparator.comparing(ConversationEntry::dateMillis).reversed());

            SwingUtilities.invokeLater(() -> {
                if (tableModel == null) return;
                tableModel.setEntries(entries);
                updateSummary(entries);
                table.getEmptyText().setText("No conversations found");
                table.getEmptyText().appendSecondaryText(
                    "Start a chat to create conversation history",
                    SimpleTextAttributes.GRAYED_ATTRIBUTES, null);
            });
        });
    }

    private List<ConversationEntry> scanConversations() {
        List<ConversationEntry> entries = new ArrayList<>();
        String basePath = project.getBasePath();
        if (basePath == null) return entries;

        Path agentWorkDir = Path.of(basePath, AGENT_WORK_DIR);
        Path currentSessionFile = agentWorkDir.resolve(CURRENT_SESSION_FILE);
        if (Files.isRegularFile(currentSessionFile)) {
            entries.add(buildEntry(currentSessionFile, true));
        }

        Path conversationsDir = agentWorkDir.resolve(CONVERSATIONS_DIR);
        if (Files.isDirectory(conversationsDir)) {
            try (DirectoryStream<Path> stream = Files.newDirectoryStream(conversationsDir,
                ARCHIVE_PREFIX + "*" + JSON_EXTENSION)) {
                for (Path file : stream) {
                    if (Files.isRegularFile(file)) {
                        entries.add(buildEntry(file, false));
                    }
                }
            } catch (IOException e) {
                LOG.warn("Failed to scan conversations directory: " + conversationsDir, e);
            }
        }

        return entries;
    }

    private static ConversationEntry buildEntry(Path file, boolean currentSession) {
        long size;
        long lastModifiedMillis;
        try {
            size = Files.size(file);
            lastModifiedMillis = Files.getLastModifiedTime(file).toMillis();
        } catch (IOException e) {
            LOG.warn("Failed to read file attributes: " + file, e);
            size = 0;
            lastModifiedMillis = 0;
        }

        int messageCount = countMessages(file);
        String displayName;
        long dateMillis;

        if (currentSession) {
            displayName = CURRENT_SESSION_LABEL;
            dateMillis = lastModifiedMillis;
        } else {
            String fileName = file.getFileName().toString();
            String timestamp = fileName
                .replace(ARCHIVE_PREFIX, "")
                .replace(JSON_EXTENSION, "");
            displayName = formatTimestamp(timestamp);
            dateMillis = parseTimestampMillis(timestamp, lastModifiedMillis);
        }

        return new ConversationEntry(file, displayName, messageCount, size, dateMillis, currentSession);
    }

    private static int countMessages(Path file) {
        try {
            String content = Files.readString(file);
            JsonArray array = JsonParser.parseString(content).getAsJsonArray();
            return array.size();
        } catch (Exception e) {
            return -1;
        }
    }

    private static String formatTimestamp(String timestamp) {
        try {
            LocalDateTime dateTime = LocalDateTime.parse(timestamp, TIMESTAMP_PARSER);
            return dateTime.format(DISPLAY_FORMATTER);
        } catch (DateTimeParseException e) {
            return timestamp;
        }
    }

    private static long parseTimestampMillis(String timestamp, long fallback) {
        try {
            LocalDateTime dateTime = LocalDateTime.parse(timestamp, TIMESTAMP_PARSER);
            return dateTime.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli();
        } catch (DateTimeParseException e) {
            return fallback;
        }
    }

    private void updateSummary(List<ConversationEntry> entries) {
        if (summaryLabel == null) return;
        if (entries.isEmpty()) {
            summaryLabel.setText(" ");
            return;
        }
        long totalSize = 0;
        for (ConversationEntry entry : entries) {
            totalSize += entry.size();
        }
        int count = entries.size();
        String countText = count == 1 ? "1 conversation" : count + " conversations";
        summaryLabel.setText(countText + " using " + formatFileSize(totalSize));
    }

    // ── Formatting utilities ──

    static String formatFileSize(long bytes) {
        if (bytes < KB) return bytes + " B";
        if (bytes < MB) return String.format("%.1f KB", bytes / (double) KB);
        return String.format("%.1f MB", bytes / (double) MB);
    }

    static String formatDateMillis(long millis) {
        if (millis <= 0) return "\u2014";
        LocalDateTime dateTime = LocalDateTime.ofInstant(
            Instant.ofEpochMilli(millis), ZoneId.systemDefault());
        return dateTime.format(DISPLAY_FORMATTER);
    }

    // ── Data record ──

    record ConversationEntry(
        Path path,
        String displayName,
        int messageCount,
        long size,
        long dateMillis,
        boolean isCurrentSession
    ) {
    }

    // ── Table model ──

    private static final class ConversationTableModel extends AbstractTableModel {
        private static final String[] COLUMNS = {"Conversation", "Messages", "Size", "Date"};
        private final List<ConversationEntry> entries = new ArrayList<>();

        @Override
        public int getRowCount() {
            return entries.size();
        }

        @Override
        public int getColumnCount() {
            return COLUMNS.length;
        }

        @Override
        public String getColumnName(int column) {
            return COLUMNS[column];
        }

        @Override
        public Object getValueAt(int rowIndex, int columnIndex) {
            ConversationEntry entry = entries.get(rowIndex);
            return switch (columnIndex) {
                case 0 -> entry.displayName();
                case 1 -> entry.messageCount() >= 0 ? String.valueOf(entry.messageCount()) : "\u2014";
                case 2 -> formatFileSize(entry.size());
                case 3 -> formatDateMillis(entry.dateMillis());
                default -> "";
            };
        }

        @Override
        public boolean isCellEditable(int rowIndex, int columnIndex) {
            return false;
        }

        ConversationEntry getEntryAt(int row) {
            return entries.get(row);
        }

        List<ConversationEntry> getEntries() {
            return List.copyOf(entries);
        }

        void setEntries(List<ConversationEntry> newEntries) {
            entries.clear();
            entries.addAll(newEntries);
            fireTableDataChanged();
        }
    }

    // ── Cell renderers ──

    private static final class ConversationNameRenderer extends ColoredTableCellRenderer {
        @Override
        protected void customizeCellRenderer(
            @NotNull JTable table, @Nullable Object value,
            boolean selected, boolean hasFocus, int row, int column
        ) {
            if (value == null) return;
            ConversationTableModel model = (ConversationTableModel) ((JBTable) table).getModel();
            ConversationEntry entry = model.getEntryAt(row);

            if (entry.isCurrentSession()) {
                setIcon(AllIcons.Actions.Execute);
                append(value.toString(), SimpleTextAttributes.REGULAR_BOLD_ATTRIBUTES);
                append("  (active)", SimpleTextAttributes.GRAYED_ATTRIBUTES);
            } else {
                setIcon(AllIcons.Vcs.History);
                append(value.toString(), SimpleTextAttributes.REGULAR_ATTRIBUTES);
            }
        }
    }

    private static final class MessageCountRenderer extends ColoredTableCellRenderer {
        @Override
        protected void customizeCellRenderer(
            @NotNull JTable table, @Nullable Object value,
            boolean selected, boolean hasFocus, int row, int column
        ) {
            setTextAlign(SwingConstants.RIGHT);
            if (value == null) return;
            String text = value.toString();
            if ("\u2014".equals(text)) {
                append(text, SimpleTextAttributes.GRAYED_ATTRIBUTES);
                setToolTipText("Unable to read message count from file");
            } else {
                append(text, SimpleTextAttributes.REGULAR_ATTRIBUTES);
                setToolTipText(null);
            }
        }
    }
}

File: plugin-core/src/main/java/com/github/catatafishen/agentbridge/settings/ChatHistoryConfigurable.kt
package com.github.catatafishen.agentbridge.settings

import com.github.catatafishen.agentbridge.services.ActiveAgentManager
import com.github.catatafishen.agentbridge.session.exporters.ExportUtils
import com.google.gson.JsonParser
import com.intellij.icons.AllIcons
import com.intellij.ide.actions.RevealFileAction
import com.intellij.notification.NotificationGroupManager
import com.intellij.notification.NotificationType
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.options.BoundConfigurable
import com.intellij.openapi.options.SearchableConfigurable
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.ui.ColoredTableCellRenderer
import com.intellij.ui.SimpleTextAttributes
import com.intellij.ui.ToolbarDecorator
import com.intellij.ui.components.JBLabel
import com.intellij.ui.dsl.builder.*
import com.intellij.ui.table.JBTable
import com.intellij.util.ui.JBUI
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import javax.swing.JTable
import javax.swing.ListSelectionModel
import javax.swing.SwingConstants
import javax.swing.table.AbstractTableModel

class ChatHistoryConfigurable(private val project: Project) :
    BoundConfigurable("Chat History"),
    SearchableConfigurable {

    override fun getId(): String = ID

    private val summaryLabel = JBLabel().apply { border = JBUI.Borders.emptyBottom(4) }
    private val tableModel = ConversationTableModel()
    private val table = JBTable(tableModel).apply {
        setSelectionMode(ListSelectionModel.MULTIPLE_INTERVAL_SELECTION)
        emptyText.text = "Start a chat to create conversation history"
        accessibleContext.accessibleName = "Conversation history files"
        columnModel.getColumn(0).preferredWidth = JBUI.scale(200)
        columnModel.getColumn(1).preferredWidth = JBUI.scale(80)
        columnModel.getColumn(2).preferredWidth = JBUI.scale(80)
        columnModel.getColumn(3).preferredWidth = JBUI.scale(160)
        columnModel.getColumn(0).cellRenderer = ConversationNameRenderer()
        columnModel.getColumn(1).cellRenderer = MessageCountRenderer()
        columnModel.getColumn(2).cellRenderer = object : ColoredTableCellRenderer() {
            override fun customizeCellRenderer(
                table: JTable, value: Any?,
                selected: Boolean, hasFocus: Boolean, row: Int, column: Int
            ) {
                if (value != null) append(value.toString(), SimpleTextAttributes.REGULAR_ATTRIBUTES)
                setTextAlign(SwingConstants.RIGHT)
            }
        }
    }

    override fun createPanel() = panel {
        val histSettings = ChatHistorySettings.getInstance(project)

        row {
            comment(
                "Conversation sessions stored in the AgentBridge sessions directory."
            )
        }
        row { cell(summaryLabel) }
        row {
            checkBox("Inject prior conversation summary into new sessions")
                .comment(
                    "When enabled, a compressed summary of recent conversation history " +
                        "is prepended to the first prompt of each new session. " +
                        "Useful as a fallback for clients without native session restore, " +
                        "but uses extra tokens. Disabled by default."
                )
                .bindSelected(
                    { ActiveAgentManager.getInjectConversationHistory(project) },
                    { ActiveAgentManager.setInjectConversationHistory(project, it) }
                )
        }
        group("History Limits") {
            row("Web event log size:") {
                spinner(100..10_000, 100)
                    .comment("Maximum number of JS events buffered for web/PWA clients")
                    .bindIntValue({ histSettings.eventLogSize }, { histSettings.eventLogSize = it })
            }
            row("DOM message limit:") {
                spinner(10..1000, 10)
                    .comment("Maximum chat messages visible in the DOM before older ones are trimmed")
                    .bindIntValue(
                        { histSettings.domMessageLimit },
                        { histSettings.domMessageLimit = it }
                    )
            }
            row("Recent turns on restore:") {
                spinner(1..100, 1)
                    .comment("Number of recent turns loaded immediately when restoring a session")
                    .bindIntValue(
                        { histSettings.recentTurnsOnRestore },
                        { histSettings.recentTurnsOnRestore = it }
                    )
            }
            row("Load-more batch size:") {
                spinner(1..50, 1)
                    .comment("Number of turns loaded per 'Load More' click")
                    .bindIntValue(
                        { histSettings.loadMoreBatchSize },
                        { histSettings.loadMoreBatchSize = it }
                    )
            }
        }
        row {
            val decorated = ToolbarDecorator.createDecorator(table)
                .disableAddAction()
                .disableUpDownActions()
                .setRemoveAction { deleteSelectedConversations() }
                .setRemoveActionUpdater {
                    val rows = table.selectedRows
                    rows.isNotEmpty() && rows.none { tableModel.getEntryAt(it).isCurrentSession }
                }
                .addExtraAction(createDeleteAllArchivesAction())
                .addExtraAction(createRefreshAction())
                .addExtraAction(createRevealInFinderAction())
                .createPanel()
            cell(decorated)
                .align(AlignX.FILL).align(AlignY.FILL).resizableColumn()
                .onReset { loadConversations() }
        }.resizableRow().layout(RowLayout.PARENT_GRID)

        ApplicationManager.getApplication().invokeLater(::loadConversations)
    }

    private fun createDeleteAllArchivesAction(): AnAction =
        object : AnAction(
            "Delete All Archives",
            "Delete all archived conversations (keeps current session)",
            AllIcons.Actions.GC
        ) {
            override fun actionPerformed(e: AnActionEvent) {
                val result = Messages.showYesNoDialog(
                    project,
                    "Delete all archived conversations?\nThe current session will be kept.\n\n" +
                        "This action cannot be undone.",
                    "Delete All Archives",
                    Messages.getWarningIcon()
                )
                if (result != Messages.YES) return
                val toDelete = tableModel.entries
                    .filter { !it.isCurrentSession }
                    .map { it.path }
                deleteFiles(toDelete)
                loadConversations()
            }

            override fun getActionUpdateThread(): ActionUpdateThread = ActionUpdateThread.EDT

            override fun update(e: AnActionEvent) {
                val hasArchives = tableModel.entries.any { !it.isCurrentSession }
                e.presentation.isEnabled = hasArchives
            }
        }

    private fun createRefreshAction(): AnAction =
        object : AnAction("Refresh", "Rescan conversation directory", AllIcons.Actions.Refresh) {
            override fun actionPerformed(e: AnActionEvent) = loadConversations()
            override fun getActionUpdateThread(): ActionUpdateThread = ActionUpdateThread.EDT
        }

    private fun createRevealInFinderAction(): AnAction =
        object : AnAction(
            "Show in ${RevealFileAction.getFileManagerName()}",
            "Open sessions directory in file manager",
            AllIcons.Actions.MenuOpen
        ) {
            override fun actionPerformed(e: AnActionEvent) {
                RevealFileAction.openDirectory(ExportUtils.sessionsDir(project))
            }

            override fun getActionUpdateThread(): ActionUpdateThread = ActionUpdateThread.EDT
        }

    private fun deleteSelectedConversations() {
        val selectedRows = table.selectedRows
        if (selectedRows.isEmpty()) return
        val message = if (selectedRows.size == 1)
            "Delete the selected conversation?\n\nThis action cannot be undone."
        else
            "Delete ${selectedRows.size} selected conversations?\n\nThis action cannot be undone."
        val result = Messages.showYesNoDialog(
            project, message, "Delete Conversations", Messages.getWarningIcon()
        )
        if (result != Messages.YES) return
        val toDelete = selectedRows.map { tableModel.getEntryAt(it).path }
        deleteFiles(toDelete)
        loadConversations()
    }

    private fun deleteFiles(paths: List<Path>) {
        val failures = mutableListOf<String>()
        for (path in paths) {
            try {
                Files.deleteIfExists(path)
            } catch (e: IOException) {
                LOG.warn("Failed to delete conversation file: $path", e)
                failures.add(path.fileName.toString())
            }
        }
        if (failures.isNotEmpty()) {
            NotificationGroupManager.getInstance()
                .getNotificationGroup("AgentBridge Notifications")
                .createNotification(
                    "Failed to delete ${failures.size} file(s): ${failures.joinToString(", ")}",
                    NotificationType.WARNING
                )
                .notify(project)
        }
    }

    private fun loadConversations() {
        table.emptyText.text = "Loading…"
        tableModel.setEntries(emptyList())
        updateSummary(emptyList())
        ApplicationManager.getApplication().executeOnPooledThread {
            val entries = scanConversations().sortedByDescending { it.dateMillis }
            ApplicationManager.getApplication().invokeLater {
                tableModel.setEntries(entries)
                updateSummary(entries)
                table.emptyText.text = "No conversations found"
                table.emptyText.appendSecondaryText(
                    "Start a chat to create conversation history",
                    SimpleTextAttributes.GRAYED_ATTRIBUTES, null
                )
            }
        }
    }

    private fun scanConversations(): List<ConversationEntry> {
        val sessionsDir = ExportUtils.sessionsDir(project).toPath()
        if (!Files.isDirectory(sessionsDir)) return emptyList()
        val currentSessionId = runCatching {
            Files.readString(sessionsDir.resolve(".current-session-id")).trim()
        }.getOrNull()
        val indexFile = sessionsDir.resolve("sessions-index.json")
        return if (Files.isRegularFile(indexFile))
            scanFromIndex(sessionsDir, indexFile, currentSessionId)
        else
            scanFromDirectory(sessionsDir, currentSessionId)
    }

    private fun scanFromIndex(sessionsDir: Path, indexFile: Path, currentSessionId: String?): List<ConversationEntry> {
        val normalizedDir = sessionsDir.normalize()
        return try {
            val array = JsonParser.parseString(Files.readString(indexFile)).asJsonArray
            val entries = mutableListOf<ConversationEntry>()
            for (el in array) {
                try {
                    val obj = el.asJsonObject
                    val id = obj.get("id")?.asString ?: continue
                    val jsonlPath = obj.get("jsonlPath")?.asString ?: "$id.jsonl"
                    val updatedAt = obj.get("updatedAt")?.asLong ?: 0L
                    val jsonlFile = sessionsDir.resolve(jsonlPath).normalize()
                    if (!jsonlFile.startsWith(normalizedDir)) continue  // path traversal guard
                    if (!Files.isRegularFile(jsonlFile)) continue
                    val size = runCatching { Files.size(jsonlFile) }.getOrDefault(0L)
                    val turnCount = obj.get("turnCount")?.takeIf { !it.isJsonNull }?.asInt
                    val messageCount = turnCount ?: ConversationFileUtils.countJsonlLines(jsonlFile)
                    val isCurrentSession = id == currentSessionId
                    val displayName = if (isCurrentSession) CURRENT_SESSION_LABEL
                    else ConversationFileUtils.formatDateMillis(updatedAt)
                    entries.add(
                        ConversationEntry(
                            jsonlFile,
                            displayName,
                            messageCount,
                            size,
                            updatedAt,
                            isCurrentSession
                        )
                    )
                } catch (e: Exception) {
                    LOG.warn("Skipping malformed session index entry: $el", e)
                }
            }
            entries
        } catch (e: Exception) {
            LOG.warn("Failed to parse sessions index: $indexFile", e)
            emptyList()
        }
    }

    private fun scanFromDirectory(sessionsDir: Path, currentSessionId: String?): List<ConversationEntry> {
        return runCatching {
            Files.list(sessionsDir).use { stream ->
                stream.filter { Files.isRegularFile(it) && it.fileName.toString().endsWith(".jsonl") }
                    .map { file ->
                        val id = file.fileName.toString().removeSuffix(".jsonl")
                        val size = runCatching { Files.size(file) }.getOrDefault(0L)
                        val dateMillis = runCatching { Files.getLastModifiedTime(file).toMillis() }.getOrDefault(0L)
                        val messageCount = ConversationFileUtils.countJsonlLines(file)
                        val isCurrentSession = id == currentSessionId
                        val displayName = if (isCurrentSession) CURRENT_SESSION_LABEL
                        else ConversationFileUtils.formatDateMillis(dateMillis)
                        ConversationEntry(file, displayName, messageCount, size, dateMillis, isCurrentSession)
                    }
                    .sorted(Comparator.comparingLong<ConversationEntry> { it.dateMillis }.reversed())
                    .toList()
            }
        }.getOrElse { e ->
            LOG.warn("Failed to scan sessions directory: $sessionsDir", e)
            emptyList()
        }
    }

    private fun updateSummary(entries: List<ConversationEntry>) {
        if (entries.isEmpty()) {
            summaryLabel.text = " "
            return
        }
        val totalSize = entries.sumOf { it.size }
        val countText = if (entries.size == 1) "1 conversation" else "${entries.size} conversations"
        summaryLabel.text = "$countText using ${ConversationFileUtils.formatFileSize(totalSize)}"
    }

    @JvmRecord
    data class ConversationEntry(
        val path: Path,
        val displayName: String,
        val messageCount: Int,
        val size: Long,
        val dateMillis: Long,
        val isCurrentSession: Boolean
    )

    private class ConversationTableModel : AbstractTableModel() {
        private val rows = mutableListOf<ConversationEntry>()
        val entries: List<ConversationEntry> get() = rows.toList()

        override fun getRowCount(): Int = rows.size
        override fun getColumnCount(): Int = COLUMNS.size
        override fun getColumnName(column: Int): String = COLUMNS[column]
        override fun getValueAt(rowIndex: Int, columnIndex: Int): Any {
            val entry = rows[rowIndex]
            return when (columnIndex) {
                0 -> entry.displayName
                1 -> if (entry.messageCount >= 0) entry.messageCount.toString() else "—"
                2 -> ConversationFileUtils.formatFileSize(entry.size)
                3 -> ConversationFileUtils.formatDateMillis(entry.dateMillis)
                else -> ""
            }
        }

        fun getEntryAt(row: Int): ConversationEntry = rows[row]

        fun setEntries(newEntries: List<ConversationEntry>) {
            rows.clear()
            rows.addAll(newEntries)
            fireTableDataChanged()
        }

        companion object {
            private val COLUMNS = arrayOf("Conversation", "Messages", "Size", "Date")
        }
    }

    private class ConversationNameRenderer : ColoredTableCellRenderer() {
        override fun customizeCellRenderer(
            table: JTable, value: Any?,
            selected: Boolean, hasFocus: Boolean, row: Int, column: Int
        ) {
            if (value == null) return
            val model = table.model as ConversationTableModel
            val entry = model.getEntryAt(row)
            if (entry.isCurrentSession) {
                icon = AllIcons.Actions.Execute
                append(value.toString(), SimpleTextAttributes.REGULAR_BOLD_ATTRIBUTES)
                append("  (active)", SimpleTextAttributes.GRAYED_ATTRIBUTES)
            } else {
                icon = AllIcons.Vcs.History
                append(value.toString(), SimpleTextAttributes.REGULAR_ATTRIBUTES)
            }
        }
    }

    private class MessageCountRenderer : ColoredTableCellRenderer() {
        override fun customizeCellRenderer(
            table: JTable, value: Any?,
            selected: Boolean, hasFocus: Boolean, row: Int, column: Int
        ) {
            setTextAlign(SwingConstants.RIGHT)
            if (value == null) return
            val text = value.toString()
            if (text == "—") {
                append(text, SimpleTextAttributes.GRAYED_ATTRIBUTES)
                toolTipText = "Unable to read message count from file"
            } else {
                append(text, SimpleTextAttributes.REGULAR_ATTRIBUTES)
                toolTipText = null
            }
        }
    }

    companion object {
        const val ID = "com.github.catatafishen.agentbridge.chatHistory"
        private val LOG = Logger.getInstance(ChatHistoryConfigurable::class.java)
        private const val CURRENT_SESSION_LABEL = "Current Session"
    }
}
